### DISCUSSION TITLE
Shutdown after export

### DISCUSSION THREAD

**Discussion by Se7enair** [RhetTbull/osxphotos#719] [Q&A]
As long as this wonderful piece of software is not running on linux, I have to start the mac and shut it down in the middle of the night to export my photos on a daily base. 
So far everything is running fine, except the shutdown.

I made a short automation on the mac which runs osxphotos. But how can I shutdown the mac when the export is finished? A simple apple script will shut down immediately.

Any ideas how to check if osxphotos is finished and then shutdown the mac?

**Comment by RhetTbull** [comment]
You could run osxphotos from a shell script that calls shutdown after osxphotos: `shutdown -h now`

If you're using the Mac only to export photos then maybe [iCloud Photos Downloader](https://github.com/icloud-photos-downloader/icloud_photos_downloader) would be a better tool?  I have no plans to port osxphotos to other operating systems as it uses a lot of Mac only features to work.

**Reply by Se7enair** [reply]
I didnt get a shell script running. At the moment I am using a apple script.
Can you tell me what I have to do? Apple automator is completely new for me....
![image](https://user-images.githubusercontent.com/1680106/174879185-8ec833e2-2f1b-43a0-ba39-29bb9c220515.png)

I looked at iCloud Photos Downloader, first it was the better solution, because it runs on linux. But there is no option to save pictures in albums and without albums.
This is not possible as fas as I know
![image](https://user-images.githubusercontent.com/1680106/174880200-0ce65c95-4272-4958-a368-fb8c62f22a1e.png)

And I am not sure if there is a option like `cleanup` and `exiftool`

But the mac is virtual and for now its better then nothing :smile:

**Reply by RhetTbull** [reply]
You can add this to your apple script to shutdown:

```applescript
tell application "System Events"
	
	shutdown
	
end tell
```

**Comment by Se7enair** [comment]
I found the way with a shell script. Problem was to tell the shell where osxphotos is.

Now I did two test runs. The mac is booting, the automator is running the script and shut down the mac afterwards.
So I got what I was trying to achieve. Fully automated backup of all iCloud photos and export of all metadata.

**Reply by RhetTbull** [reply]
Are you willing to share more details of your setup? For example, how you configured the Mac VM, your scripts, etc. It might be helpful to others who want to try a similar workflow.

**Comment by Se7enair** [comment]
If course, yes. That was the plan. Will write a short tutorial in the next days.

**Comment by Se7enair** [comment]
## **How to automate a backup mit osxphotos and a Mac:**

I used osxphotos and apple automator to get a complete automated backup of all my photos which are saved in the iCloud.
I am doing this because I want a backup on my local server, just in case. And to have the possibility to change to another photo service at any time, without loosing any album or other metadata like positions, detected persons, etc.
With this, you have to do all changes via Apple Photos or your iOS device, because every change on the local files will be overwritten. As I said, at the moment it is only a backup. Maybe you can use it as a read only source for eg Plex. 


### **Preparations:**
**Grant sudo access to shutdown without password:**
My way destroys sudo. It works, but after I cannot do any commands with sudo. Have to repair and find a better way. But you have to do it, otherwise the script will stop, because shutdown cannot be executed without sudo. 

**Create config file for osxphotos:**
You need a config-file, or you pass all options directly in the shell-script.
```
  --save-config CONFIG_FILE       Save options to file for use with --load-
                                  config. File format is TOML. See also
                                  --config-only.
  --config-only                   If specified, saves the config file but does
                                  not export any files; must be used with
                                  --save-config.
```
This is mine for example:
```
[export]
album_keyword = true
cleanup = true
directory = "{created.year}/{created.mm}/{album,}"
edited_suffix = "_EDIT_{modified.year}{modified.mm}{modified.dd}"
exiftool = true
exiftool_path = "/usr/local/bin/exiftool"
filename_template = "{created.date}_{created.hour}{created.min}{created.sec}"
finder_tag_keywords = true
force_update = true
from_date = 2018-01-01T00:00:00
jpeg_ext = "jpg"
person_keyword = true
verbose = true
db = "/Users/christoph/Pictures/Photos Library.photoslibrary/database/photos.db"
download_missing = true
dry_run = false
```

**1. Automatic start of the Mac**
Because I am using a virtual Mac, the start is not suitable for the must users. I you use a real Mac, under settings there is a option to schedule a startup. I did not tested this.
![Bildschirmfoto 2022-06-22 um 21 47 23](https://user-images.githubusercontent.com/1680106/175123546-6c92d340-bf29-4236-b6cf-0c0fe7bcdb9b.png)

**2. Let Automator start your script and shut down the Mac**
Open automator and create a new program. Add a shell script.
![Bildschirmfoto 2022-06-22 um 21 50 34](https://user-images.githubusercontent.com/1680106/175129040-554ea477-eb2f-457b-accf-2f47fc777116.png)
`/Users/your username/.local/bin/osxphotos export /path/to/export --load-configpath/to/your/config
sudo shutdown -h now`
You can adapt this to your needs. You can for example delete the load-config option and write down all options you need.
Note that I used `exiftool-path`, without the shell-script will abort, because it cannot find exiftool.

save it in your preferred directory

**3. Schedule the automator programm**
Open the calendar app on your Mac. Create a new event. For example: event is repeated everyday at 00:10.
Under alarms choose "other" and then "open file". After that search for your automator Programm and choose it.
![Bildschirmfoto 2022-06-22 um 21 59 13](https://user-images.githubusercontent.com/1680106/175125353-4be4f127-d403-46f4-b57c-6007c7b0746c.png)

Finished. Your Mac should now boot at your scheduled time, run the script and shut down afterwards. Keep in mind that the Mac will shut down if you maybe working on it. I use the Mac only for this, so for me this is fine.

**Reply by RhetTbull** [reply]
@Se7enair thank you for the detailed steps -- this is very helpful!  Do you have any links on how to set up the virtual machine and the virtual Mac?  Are you using Parallels, Virtual Box, etc?  Did you need to do anything special for the virtual Mac to be able to connect to iCloud (I've never tried a virtual Mac but I've heard this can be an issue).

**Comment by Se7enair** [comment]
I fact, it was easier then a hackintosh.
I use proxmox as a hypervisor und this [tutorial ](https://www.nicksherlock.com/2021/10/installing-macos-12-monterey-on-proxmox-7/)to set the mac up. You need a real mac to fetch a OSK authentication key. But every mac uses the same OSK, so don't be surprised that it doesn't look like a random string.
So far I have no problems with iCloud at all. The virtual machine is started via cronjob in proxmox, just put the command `qm start <vmid>` as a cronjob. At the moment I cannot show mine, because I have no access to the hypervisor. But for someone who is familiar with linux, this should not be a problem.

**Reply by RhetTbull** [reply]
Thanks!

**Comment by Se7enair** [comment]
In proxmox go to the shell of the node where your vm lives. Type 

`nano /etc/crontab`

add this to your other cronjobs

`55 23   * * *   root    qm start <vmid>`

exit and save

then type 

`systemctl restart cron.service`

and the vm should start every day at 23:55

**Comment by RhetTbull** [comment]
@Se7enair osxphotos now runs on linux which may simplify your workflow. (Thanks to the work of @dvdkon who did the port)

**Reply by Se7enair** [reply]
Wow, Sounds good. 
Will try if i find some time. My solution with a virtual mac never was really stable.
Fun fact: i actually looking forward to buy a mac :D

**Reply by Se7enair** [reply]
Do you have any tips how to get the photos library to a linux computer? I don't want/cannot sync from a mac.

**Reply by RhetTbull** [reply]
No, I don't have any tips on that, sorry. You could use your VM trick you were doing and keep the library on the linux host then run osxphotos from the linux host if that's more stable than your original solution.
